zebra-chain/src/transaction.rs

```diff
@@ -76,44 +76,133 @@ pub struct TransparentOutput {
+    // for now it's [u64; 8] rather than [u8; 64] to get trait impls
+    pub binding_sig: [u64; 8],
+}
 
-    /// A list of 1 or more transparent transaction inputs or sources
-    /// for coins.
-    pub tx_in: Vec<TransparentInput>,
+/// unimplemented.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct JoinSplitBctv14 {}
+/// unimplemented.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct JoinSplitGroth16 {}
+
+/// Pre-Sapling JoinSplit data using Sprout-on-BCTV14 proofs.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct LegacyJoinSplitData {
+    /// A sequence of JoinSplit descriptions using BCTV14 proofs.
+    pub join_splits: Vec<JoinSplitBctv14>,
+    /// The public key for the JoinSplit signature.
+    // XXX refine to a Zcash-flavored Ed25519 pubkey.
+    pub pub_key: [u8; 32],
+    /// The JoinSplit signature.
+    // XXX refine to a Zcash-flavored Ed25519 signature.
+    // for now it's [u64; 8] rather than [u8; 64] to get trait impls
+    pub sig: [u64; 8],
+}
 
-    /// A list of 1 or more transparent transaction outputs or
-    /// destinations for coins.
-    pub tx_out: Vec<TransparentOutput>,
+/// Post-Sapling JoinSplit data using Sprout-on-Groth16 proofs.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SaplingJoinSplitData {
+    /// A sequence of JoinSplit descriptions using BCTV14 proofs.
+    pub join_splits: Vec<JoinSplitGroth16>,
+    /// The public key for the JoinSplit signature.
+    // XXX refine to a Zcash-flavored Ed25519 pubkey.
+    pub pub_key: [u8; 32],
+    /// The JoinSplit signature.
+    // XXX refine to a Zcash-flavored Ed25519 signature.
+    // for now it's [u64; 8] rather than [u8; 64] to get trait impls
+    pub sig: [u64; 8],
+}
 
-    /// The block number or timestamp at which this transaction is unlocked.
-    ///
-    /// If all `TransactionInput`s have final (0xffffffff) sequence
-    /// numbers, then lock_time is irrelevant. Otherwise, the
-    /// transaction may not be added to a block until after `lock_time`.
-    pub lock_time: LockTime,
-
-    /// A block height in the range {1 .. 499999999} after which the
-    /// transaction will expire, or 0 to disable expiry ([ZIP-203]).
-    pub expiry_height: u32,
+/// A Zcash transaction.
+///
+/// A transaction is an encoded data structure that facilitates the transfer of
+/// value between two public key addresses on the Zcash ecosystem. Everything is
+/// designed to ensure that transactions can created, propagated on the network,
+/// validated, and finally added to the global ledger of transactions (the
+/// blockchain).
+///
+/// Zcash has a number of different transaction formats. They are represented
+/// internally by different enum variants. Because we checkpoint on Sapling
+/// activation, we do not parse any pre-Sapling transaction types.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Transaction {
+    /// A fully transparent transaction (`version = 1`).
+    Transparent {
```

Comment:
I'd suggest naming these as V1, V2, V3, V4, since transaction versions don't correspond 1:1 with upgrade epochs &mdash; e.g. Blossom won't create a new version; pre-Overwinter txns can be either V1 or V2; V2 or later txns can be fully transparent; and so on.

Reply:
This is a much better naming scheme -- I implemented it in a later commit.